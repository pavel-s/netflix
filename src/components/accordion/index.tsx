import {
  createContext,
  Dispatch,
  SetStateAction,
  useContext,
  useState,
} from 'react';
import { TRSC } from '../../types';
import {
  Container,
  Inner,
  Body,
  Frame,
  Header,
  Item,
  Title,
} from './styles/accordion';

type TAccordion = TRSC<typeof Container> & {
  Title: TRSC<typeof Title>;
  Frame: TRSC<typeof Frame>;
  Item: TRSC<typeof Item>;
  Header: TRSC<typeof Header>;
  Body: TRSC<typeof Body>;
};

const Accordion: TAccordion = ({ children, ...rest }) => {
  return (
    <Container {...rest}>
      <Inner>{children}</Inner>
    </Container>
  );
};

const ToggleContext = createContext<{
  show?: boolean;
  setShow?: Dispatch<SetStateAction<boolean>>;
}>({});

Accordion.Title = ({ children, ...rest }) => (
  <Title {...rest}>{children}</Title>
);

Accordion.Frame = ({ children, ...rest }) => (
  <Frame {...rest}>{children}</Frame>
);

Accordion.Item = function AccordionItem({ children, ...rest }) {
  const [show, setShow] = useState(false);

  return (
    <ToggleContext.Provider value={{ show, setShow }}>
      <Item {...rest}>{children}</Item>
    </ToggleContext.Provider>
  );
};

Accordion.Header = function AccordionHeader({ children, ...rest }) {
  const { show, setShow } = useContext(ToggleContext);

  const toggleShow = () => {
    setShow && setShow((show: boolean) => !show);
  };

  return (
    <Header onClick={toggleShow} {...rest}>
      {children}{' '}
      {show ? (
        <img src='/images/icons/close-slim.png' alt='' />
      ) : (
        <img src='/images/icons/add.png' alt='' />
      )}
    </Header>
  );
};

Accordion.Body = function AccordionBody({ children, ...rest }) {
  const { show } = useContext(ToggleContext);
  return (
    <Body className={show ? 'open' : 'closed'}>
      <span {...rest}>{children}</span>
    </Body>
  );
};

export default Accordion;
