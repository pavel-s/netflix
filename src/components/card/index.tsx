import { TRSC } from '../../types';

import {
  Container,
  CardContainer,
  Group,
  Title,
  Items,
  Item,
  ItemImage,
  Meta,
  Subtitle,
  Text,
  Feature,
  Maturity,
  FeatureClose,
} from './styles/card';

type TCard = TRSC<typeof Container> & {
  Container: TRSC<typeof Container>;
  Group: TRSC<typeof Group>;
  Title: TRSC<typeof Title>;
  Items: TRSC<typeof Items>;
  Item: TRSC<typeof Item>;
  ItemImage: TRSC<typeof ItemImage>;
  Meta: TRSC<typeof Meta>;
  Subtitle: TRSC<typeof Subtitle>;
  Text: TRSC<typeof Text>;
  Feature: TRSC<typeof Feature>;
  Maturity: TRSC<typeof Maturity>;
  FeatureClose: TRSC<typeof FeatureClose>;
};

const Card: TCard = ({ children, ...rest }) => {
  return <CardContainer {...rest}>{children}</CardContainer>;
};

Card.Container = function CardCardContainer({ children, ...rest }) {
  return <Container {...rest}>{children}</Container>;
};

Card.Group = function CardGroup({ children, ...rest }) {
  return <Group {...rest}>{children}</Group>;
};

Card.Title = function CardTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Card.Items = function CardItems({ children, ...rest }) {
  return <Items {...rest}>{children}</Items>;
};

Card.Item = function CardItem({ children, ...rest }) {
  return <Item {...rest}>{children}</Item>;
};

Card.ItemImage = function CardItemImage({ children, ...rest }) {
  return <ItemImage {...rest}>{children}</ItemImage>;
};

Card.Meta = function CardMeta({ children, ...rest }) {
  return <Meta {...rest}>{children}</Meta>;
};

Card.Subtitle = function CardSubtitle({ children, ...rest }) {
  return <Subtitle {...rest}>{children}</Subtitle>;
};

Card.Text = function CardText({ children, ...rest }) {
  return <Text {...rest}>{children}</Text>;
};

Card.Feature = function CardFeature({ children, ...rest }) {
  return <Feature {...rest}>{children}</Feature>;
};

Card.Maturity = function CardMaturity({ maturity, ...rest }) {
  return (
    <Maturity maturity={maturity} {...rest}>
      {maturity}
    </Maturity>
  );
};

Card.FeatureClose = function CardFeatureClose({ children, ...rest }) {
  return <FeatureClose {...rest}>{children}</FeatureClose>;
};

export default Card;
