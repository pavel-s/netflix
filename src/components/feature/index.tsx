import { TRSC } from '../../types';
import { Container, Inner, Title, Subtitle } from './styles/feature';

type TFeature = TRSC<'div'> & {
  Title: TRSC<'h1'>;
  Subtitle: TRSC<'h2'>;
};

const Feature: TFeature = ({ children, ...rest }) => {
  return (
    <Container {...rest}>
      <Inner>{children}</Inner>
    </Container>
  );
};

Feature.Title = function FeatureTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Feature.Subtitle = function FeatureSubtitle({ children, ...rest }) {
  return <Subtitle {...rest}>{children}</Subtitle>;
};

export default Feature;
