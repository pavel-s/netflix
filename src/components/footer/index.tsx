import { TRSC } from '../../types';
import { Container, Inner, Title, Links, Link, Text } from './styles/footer';

type TFooter = TRSC<'div'> & {
  Title: TRSC<'p'>;
  Links: TRSC<'ul'>;
  Link: TRSC<'a'>;
  Text: TRSC<'p'>;
};

const Footer: TFooter = ({ children, ...rest }) => {
  return (
    <Container {...rest}>
      <Inner>{children}</Inner>
    </Container>
  );
};

Footer.Title = ({ children, ...rest }) => {
  return <Title {...rest}>{children}</Title>;
};

Footer.Links = ({ children, ...rest }) => {
  return <Links {...rest}>{children}</Links>;
};

Footer.Link = ({ children, ...rest }) => {
  return (
    <Link>
      <a {...rest}>{children}</a>
    </Link>
  );
};

Footer.Text = ({ children, ...rest }) => {
  return <Text {...rest}>{children}</Text>;
};

export default Footer;
