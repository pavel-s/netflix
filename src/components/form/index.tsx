import { TRSC } from '../../types';
import {
  Container,
  Base,
  Title,
  Input,
  Error,
  Submit,
  Row,
  Checkbox,
  Text,
  TextBig,
  Link,
  TextButton,
} from './styles/form';

type TForm = TRSC<typeof Container> & {
  Base: TRSC<typeof Base>;
  Title: TRSC<typeof Title>;
  Input: TRSC<typeof Input>;
  Error: TRSC<typeof Error>;
  Submit: TRSC<typeof Submit>;
  Row: TRSC<typeof Row>;
  Checkbox: TRSC<typeof Checkbox>;
  Text: TRSC<typeof Text>;
  TextBig: TRSC<typeof TextBig>;
  Link: TRSC<typeof Link>;
  TextButton: TRSC<typeof TextButton>;
};

const Form: TForm = ({ children, ...rest }) => {
  return <Container {...rest}>{children}</Container>;
};

Form.Base = function FormBase({ children, ...rest }) {
  return <Base {...rest}>{children}</Base>;
};

Form.Title = function FormTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Form.Input = function FormInput({ ...rest }) {
  return <Input {...rest} />;
};

Form.Error = function FormError({ ...rest }) {
  return <Error data-testid='form-error' {...rest} />;
};

Form.Submit = function FormSubmit({ children, ...rest }) {
  return (
    <Submit data-testid='form-submit' {...rest}>
      {children}
    </Submit>
  );
};

Form.Row = function FormRow({ children, ...rest }) {
  return <Row {...rest}>{children}</Row>;
};

Form.Checkbox = function FormCheckbox({ children, ...rest }) {
  return <Checkbox {...rest}>{children}</Checkbox>;
};

Form.Text = function FormText({ children, ...rest }) {
  return <Text {...rest}>{children}</Text>;
};

Form.TextBig = function FormTextBig({ children, ...rest }) {
  return <TextBig {...rest}>{children}</TextBig>;
};

Form.Link = function FormLink({ children, ...rest }) {
  return <Link {...rest}>{children}</Link>;
};

Form.TextButton = function FormTextButton({ children, ...rest }) {
  return <TextButton {...rest}>{children}</TextButton>;
};

export default Form;
