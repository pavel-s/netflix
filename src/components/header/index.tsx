import { Background, Container, Logo, SignButton } from './styles/header';
import { LinkProps } from 'react-router-dom';
import { TRSC } from '../../types';

type THeader = TRSC<'div'> & {
  // Container: TRSC<'div'>;
  Logo: TRSC<'img'>;
  Frame: TRSC<'div'>;
  SignButton: TRSC<'a', LinkProps>;
};

const Header: THeader = ({ children, ...rest }) => {
  return <Background {...rest}>{children}</Background>;
};

Header.Frame = function HeaderFrame({ children, ...rest }) {
  return <Container {...rest}>{children}</Container>;
};

Header.Logo = function HeaderLogo({ children, ...rest }) {
  return <Logo {...rest} />;
};

Header.SignButton = function HeaderSignButton({ children, ...rest }) {
  return <SignButton {...rest}>{children}</SignButton>;
};

export default Header;
