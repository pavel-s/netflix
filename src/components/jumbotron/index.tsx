import { TRSC } from '../../types';
import {
  Container,
  Item,
  Inner,
  Pane,
  Title,
  Subtitle,
  Image,
  TFlexDirection,
} from './styles/jumbotron';

type TJumbotron = TRSC<typeof Item, { direction: TFlexDirection }> & {
  Container: TRSC<typeof Container>;
  Pane: TRSC<typeof Pane>;
  Title: TRSC<typeof Title>;
  Subtitle: TRSC<typeof Subtitle>;
  Image: TRSC<typeof Image>;
};

const Jumbotron: TJumbotron = ({ direction = 'row', children, ...rest }) => {
  return (
    <Item {...rest}>
      <Inner direction={direction}>{children}</Inner>
    </Item>
  );
};

Jumbotron.Container = function JumbotronContainer({ children, ...rest }) {
  return <Container {...rest}>{children}</Container>;
};

Jumbotron.Pane = function JumbotronPane({ direction, children, ...rest }) {
  return (
    <Pane direction={direction} {...rest}>
      {children}
    </Pane>
  );
};

Jumbotron.Title = function JumbotronTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Jumbotron.Subtitle = function JumbotronSubtitle({ children, ...rest }) {
  return <Subtitle {...rest}>{children}</Subtitle>;
};

Jumbotron.Image = function JumbotronImage({ children, ...rest }) {
  return (
    <div>
      <Image {...rest} />
    </div>
  );
};

export default Jumbotron;
