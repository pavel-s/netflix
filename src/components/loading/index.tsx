import React from 'react';
import { TRSC } from '../../types';
import { Container, Image, LockBody, Spinner } from './styles/loading';

type TLoading = TRSC<
  typeof Image,
  Pick<typeof Spinner extends React.ComponentType<infer P> ? P : {}, 'size'>
>;

const Loading: TLoading = ({ size, src, ...rest }) => {
  return (
    <Container data-testid='loading'>
      <LockBody />
      <Spinner size={size}>
        {src ? <Image src={src} {...rest} /> : null}
      </Spinner>
    </Container>
  );
};

export default Loading;
