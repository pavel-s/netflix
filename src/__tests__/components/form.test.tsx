import { render } from '@testing-library/react';
import { Form } from '../../components';
import { MemoryRouter } from 'react-router-dom';

describe('<Form />', () => {
  it('renders Form with populated data', () => {
    const { container, queryByText, queryByPlaceholderText, queryByTestId } =
      render(
        <Form>
          <Form.Title>Sign In Now</Form.Title>
          <Form.Base>
            <Form.Error>Form error</Form.Error>
            <Form.Input type='text' placeholder='Email or phone number' />
            <Form.Input type='password' placeholder='Password' error={true} />

            <Form.Submit isLoading={false}>Sign In</Form.Submit>

            <Form.Row>
              <Form.Row>
                <Form.Checkbox id='rememberMe' />
                <label htmlFor='rememberMe'>Remember me</label>
              </Form.Row>
              <Form.Link to={'/home'}>Need help?</Form.Link>
            </Form.Row>
          </Form.Base>
          <Form.TextBig data-testid='text-big'>
            New to Netflix?{' '}
            <Form.Link to={'/home'} color='#fff'>
              Sign up now
            </Form.Link>
          </Form.TextBig>
          <Form.Text data-testid='text'>
            This page is protected by Google reCAPTCHA to ensure you're not a
            bot.
            <Form.TextButton>Learn more.</Form.TextButton>
          </Form.Text>
        </Form>,
        { wrapper: MemoryRouter }
      );

    // title
    expect(queryByText('Sign In Now')).toBeTruthy();

    // error
    expect(queryByText('Form error')).toBeTruthy();

    // text input
    expect(queryByPlaceholderText('Email or phone number')).toBeTruthy();

    // password input with error=true
    expect(queryByPlaceholderText('Password')).toBeTruthy();

    // submit button
    expect(queryByText('Sign In')).toHaveAttribute('type', 'submit');

    // checkbox
    expect(container.querySelector('#rememberMe')).toBeTruthy();

    // link
    expect(queryByText('Need help?')).toBeTruthy();

    // link with color
    expect(queryByText('Sign up now')).toBeTruthy();

    // big text
    expect(queryByTestId('text-big')).toBeTruthy();

    // text
    expect(queryByTestId('text')).toBeTruthy();

    // text button
    expect(queryByText('Learn more.')).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders Form with submit button in loading state', () => {
    const { container, queryByText } = render(
      <Form>
        <Form.Submit isLoading={true}>Sign In</Form.Submit>
      </Form>,
      { wrapper: MemoryRouter }
    );

    expect(queryByText('Sign In')).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });
});
