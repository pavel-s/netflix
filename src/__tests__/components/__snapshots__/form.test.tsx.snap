// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> renders Form with populated data 1`] = `
<div
  class="form__Container-sc-1e02ltb-0 dTbbCb"
>
  <h1
    class="form__Title-sc-1e02ltb-2 kkeSEH"
  >
    Sign In Now
  </h1>
  <form
    class="form__Base-sc-1e02ltb-1 NkBbX"
  >
    <p
      class="form__Error-sc-1e02ltb-4 dprvGA"
      data-testid="form-error"
    >
      Form error
    </p>
    <input
      class="form__Input-sc-1e02ltb-3 gqlhzz"
      placeholder="Email or phone number"
      type="text"
    />
    <input
      class="form__Input-sc-1e02ltb-3 gFNeLa"
      placeholder="Password"
      type="password"
    />
    <button
      class="form__Submit-sc-1e02ltb-5 fvISnT"
      data-testid="form-submit"
      type="submit"
    >
      Sign In
    </button>
    <div
      class="form__Row-sc-1e02ltb-6 ciGvba"
    >
      <div
        class="form__Row-sc-1e02ltb-6 ciGvba"
      >
        <input
          class="form__Checkbox-sc-1e02ltb-7 CTzfr"
          id="rememberMe"
          type="checkbox"
        />
        <label
          for="rememberMe"
        >
          Remember me
        </label>
      </div>
      <a
        class="form__Link-sc-1e02ltb-10 jZsvly"
        href="/home"
      >
        Need help?
      </a>
    </div>
  </form>
  <p
    class="form__TextBig-sc-1e02ltb-9 qOJIB"
    data-testid="text-big"
  >
    New to Netflix?
     
    <a
      class="form__Link-sc-1e02ltb-10 ASZx"
      color="#fff"
      href="/home"
    >
      Sign up now
    </a>
  </p>
  <p
    class="form__Text-sc-1e02ltb-8 hBXpQP"
    data-testid="text"
  >
    This page is protected by Google reCAPTCHA to ensure you're not a bot.
    <button
      class="form__TextButton-sc-1e02ltb-11 gINjTt"
    >
      Learn more.
    </button>
  </p>
</div>
`;

exports[`<Form /> renders Form with submit button in loading state 1`] = `
<div
  class="form__Container-sc-1e02ltb-0 dTbbCb"
>
  <button
    class="form__Submit-sc-1e02ltb-5 hpSdAz"
    data-testid="form-submit"
    type="submit"
  >
    Sign In
  </button>
</div>
`;
